### ISSUE TITLE
Replace antlr4 LaTeX parser?

### ISSUE THREAD

**Issue by bollwyvl** [sympy/sympy#14004]
> Probably needs some labels

If #13706 lands, there is at least a baseline for a (partially working) LaTeX parser. We can discuss a next-best (or best-of-n) replacement, as originally discussed on [latex2sympy](https://github.com/augustt198/latex2sympy/issues/1#issuecomment-349331543).

I did try out a [go at `ply`](https://github.com/bollwyvl/sympy/blob/feat/parse_latex_ply/sympy/parsing/_latex/_ply/parser.py), and feel like it is worth exploring further.

One more approach identified was [tex2py](https://github.com/alvinwan/tex2py).

**Comment by asmeurer** [comment]
CC @scopatz

**Comment by asmeurer** [comment]
Another option https://github.com/erikrose/parsimonious. An advantage is it uses a PEG parser, which is easier to deal with.

**Comment by bollwyvl** [comment]
My money would be on lark today... It's even plausible to dynamically
extend the schema.

>

**Comment by gschintgen** [comment]
I'd just like to point out that I've now lost an entire afternoon trying to get conda to install antlr4-python3-runtime (via conda-forge). Either the package is broken or it's incompatible with one of: sympy, matplotlib, seaborn, scipy or ipykernel.
I didn't expect such an "experience" given SymPy's policy/ideal of minimalistic dependencies.
Replacing antlr seems like a fine idea!

**Comment by bollwyvl** [comment]
Feel free to raise any specific conda-forge specific issues in the
respective feedstock: A quick binder suggests it's possible to get all of
those from conda-forge on at least amd64/Linux/3.7(up to 3.10), but your
mileage may vary.

We try to keep stuff running over there, but there are rather a few moving
pieces in the stack you describe on... whatever arch/os/python you might be
using. Logs would help, but sunny germane here, probably.

As to the issue at hand: LaTeX parsing has _always_ been marked an
experimental feature, and some good work has been done to try other things.
It would however, need a very dedicated volunteer to pick up the pieces
left by others to adopt a different parser, and handle the packaging,
testing, and documentation concerns. That could be you!

**Comment by gschintgen** [comment]
> Feel free to raise any specific conda-forge specific issues in the respective feedstock: A quick binder suggests it's possible to get all of those from conda-forge on at least amd64/Linux/3.7(up to 3.10), but your mileage may vary.

Thanks for your reply! At least that confirms that somehow it's my setup where things go wrong. (amd64, ubuntu 18.04 inside an LXD container, the littlest jupyterhub, rather empty conda env separate from 'base'). I'm sorry if my post was a bit too harsh.

It's not the first time conda bailed out on me while trying to solve an environment. I'm using it inside an LXD container so that I'm certain I can always recover the whole setup from a snapshot as a last resort. Conda does seem fragile sometimes when it comes to dependency resolving. But the wealth of software that is packaged is phenomenal. I'm very grateful for all the hard work that has been put in the whole ecosystem! (Even if my previous post didn't sound like it...)

**Comment by asmeurer** [comment]
I don't know if it's the cause of your problems, but one issue with the antlr parser is that the parser generator requires pinning to an exact version of the runtime (see https://github.com/conda-forge/sympy-feedstock/blob/main/recipe/meta.yaml#L30). That means that if anything else in your environment uses antlr but is based on another version, it will be impossible to install. Doing a search of conda-forge reveals a couple of packages that pin different versions of antlr-python-runtime than SymPy https://github.com/search?q=org%3Aconda-forge+antlr-python-runtime&type=code. 

By the way, I recommend using `mamba` instead of `conda`. It does a better job in instances like this, and it's also much faster.

**Comment by gschintgen** [comment]
Hm, I finally got conda to resolve the environment (a new one, but I *did* try that previously, so I'm puzzled). What's still strange is the interplay between sympy and antlr4-python3-runtime, so I'm hesitant to file an issue in the specific antlr feedstock. Excerpt from my system:
```
$ conda create -n tmp -c conda-forge sympy matplotlib seaborn ipykernel scipy
[...]
  sympy              conda-forge/linux-64::sympy-1.10-py310hff52083_0
[...]

$ conda create -n tmp -c conda-forge sympy antlr4-python3-runtime
[...]
  sympy              conda-forge/linux-64::sympy-1.5.1-py38_0
[...]
```

**Comment by gschintgen** [comment]
> I don't know if it's the cause of your problems, but one issue with the antlr parser is that the parser generator requires pinning to an exact version of the runtime (see https://github.com/conda-forge/sympy-feedstock/blob/main/recipe/meta.yaml#L30). That means that if anything else in your environment uses antlr but is based on another version, it will be impossible to install. Doing a search of conda-forge reveals a couple of packages that pin different versions of antlr-python-runtime than SymPy 

Ah, thanks, that might explain it! I guess I'll launch a throwaway container and do some more experiments. Having sympy drop to 1.5.1 as soon as antlr is added is slightly suboptimal.

**Comment by asmeurer** [comment]
Yep, it looks like 1.5.1 was the version where the runtime constraints were added to the recipe https://github.com/conda-forge/sympy-feedstock/commit/8d87ce7cd88b8a18ffc45a538cdd0488e4fa58ec, although the constraints should really be there for every version that has the latex parser because the antlr parser has always worked like this.

**Comment by gschintgen** [comment]
> Yep, it looks like 1.5.1 was the version where the runtime constraints were added to the recipe [conda-forge/sympy-feedstock@8d87ce7](https://github.com/conda-forge/sympy-feedstock/commit/8d87ce7cd88b8a18ffc45a538cdd0488e4fa58ec), although the constraints should really be there for every version that has the latex parser because the antlr parser has always worked like this.

Oh well, I think I've just found the culprit:
`antlr4-python3-runtime` is a stale old package that must have shown up on various google searches and the issue tracker here, but the right one is `antlr-python-runtime`.  Quite a pitfall.

That's a few lost hours for each of the two characters  :-O
Thanks for your responses @bollwyvl and @asmeurer !

**Comment by asmeurer** [comment]
I think `antlr4-python3-runtime` is the name of the package on PyPI. 

This whole mess is a pretty strong reason for ditching antlr in my opinion. I'd prefer a package that's much simpler as a dependency, even if it is just an optional one.
